src/libAtomVM/term.c

```diff
@@ -379,3 +379,57 @@ int term_compare(term t, term other, Context *ctx)
 
     return result;
 }
+
+term term_alloc_refc_binary(Context *ctx, const void *data, size_t size, unsigned is_const)
+{
+    term *boxed_value = memory_heap_alloc(ctx, TERM_BOXED_REFC_BINARY_SIZE);
+    boxed_value[0] = ((TERM_BOXED_REFC_BINARY_SIZE - 1) << 6) | TERM_BOXED_REFC_BINARY;
+    boxed_value[1] = (term) size;
+    boxed_value[2] = (term) is_const ? 0x01 : 0x00;
+    term ret = ((term) boxed_value) | TERM_BOXED_VALUE_TAG;
+    if (is_const) {
+        boxed_value[3] = (term) data;
+        boxed_value[4] = term_nil(); // mso_list is not used
+        boxed_value[5] = term_nil(); // for const binaries
+    } else {
+        void *refc = memory_create_refc(size, data);
+        boxed_value[3] = (term) refc;
+        ctx->mso_list = term_list_init_prepend(boxed_value + 4, ret, ctx->mso_list);
+        list_append(&ctx->global->refc_binaries, (struct ListHead *) refc);
+    }
+    return ret;
+}
+
+static term find_finary(term binary_or_state) {
+    if (term_is_match_state(binary_or_state)) {
+        find_finary(term_get_match_state_binary(binary_or_state));
+    } else {
+        return binary_or_state;
+    }
+}
+
+term term_alloc_bin_match_state(term binary_or_state, Context *ctx)
+{
+    term *boxed_match_state = memory_heap_alloc(ctx, TERM_BOXED_BIN_MATCH_STATE_SIZE);
+
+    boxed_match_state[0] = ((TERM_BOXED_BIN_MATCH_STATE_SIZE - 1) << 6) | TERM_BOXED_BIN_MATCH_STATE;
+    if (term_is_match_state(binary_or_state)) {
+        boxed_match_state[1] = term_get_match_state_binary(binary_or_state);
+        boxed_match_state[2] = term_get_match_state_offset(binary_or_state);
+        boxed_match_state[3] = term_get_match_state_offset(binary_or_state);
+    } else {
+        boxed_match_state[1] = binary_or_state;
+        boxed_match_state[2] = 0;
+        boxed_match_state[3] = 0;
+    }
+    term binary = find_finary(binary_or_state);
+    term *boxed_value = term_to_term_ptr(binary);
+    if (term_is_refc_binary(binary) && !term_refc_binary_is_const(binary)) {
+        memory_increment_recf_refcount(boxed_value[3]);
+        ctx->mso_list = term_list_init_prepend(boxed_match_state + 4, binary, ctx->mso_list);
```

Comment:
Dangerous in-memory layout hard-coding here.

Reply:
I'm not really sure what the alternative is.  We need _some_ offset to find the cons cell for the mso list.  Suggestions?

Reply:
#defining it or creating an inline function would be good enough. Maybe `#define` is the right choice.

Reply:
Let me know if that solves it.